synapse/storage/databases/state/bg_updates.py

```diff
@@ -93,158 +93,104 @@ def _get_state_groups_from_groups_txn(
+                    for state_key in state_keys:
+                        state_filter_condition_combos.append((etype, state_key))
+        # And here is the optimization itself. We don't want to do the optimization
+        # if there are too many individual conditions. 10 is an arbitrary number
+        # with no testing behind it but we do know that we specifically made this
+        # optimization for when we grab the necessary state out for
+        # `filter_events_for_client` which just uses 2 conditions
+        # (`EventTypes.RoomHistoryVisibility` and `EventTypes.Member`).
+        if use_condition_optimization and len(state_filter_condition_combos) < 10:
+            select_clause_list: List[str] = []
+            for etype, skey in state_filter_condition_combos:
+                if skey is None:
+                    where_clause = "(type = ?)"
+                    overall_select_query_args.extend([etype])
+                else:
+                    where_clause = "(type = ? AND state_key = ?)"
+                    overall_select_query_args.extend([etype, skey])
+
+                select_clause_list.append(
+                    f"""
+                    (
+                        SELECT DISTINCT ON (type, state_key)
+                            type, state_key, event_id
+                        FROM state_groups_state
+                        INNER JOIN sgs USING (state_group)
+                        WHERE {where_clause}
+                        ORDER BY type, state_key, state_group DESC
                     )
+                    """
+                )
 
-                overall_select_clause = " UNION ".join(select_clause_list)
-            else:
-                where_clause, where_args = state_filter.make_sql_filter_clause()
-                # Unless the filter clause is empty, we're going to append it after an
-                # existing where clause
-                if where_clause:
-                    where_clause = " AND (%s)" % (where_clause,)
-
-                overall_select_query_args.extend(where_args)
-
-                overall_select_clause = f"""
-                    SELECT DISTINCT ON (type, state_key)
-                        type, state_key, event_id
-                    FROM state_groups_state
-                    WHERE state_group IN (
-                        SELECT state_group FROM sgs
-                    ) {where_clause}
-                    ORDER BY type, state_key, state_group DESC
-                """
-
-            for group in groups:
-                args: List[Union[int, str]] = [group]
-                args.extend(overall_select_query_args)
-
-                txn.execute(sql % (overall_select_clause,), args)
-                for row in txn:
-                    typ, state_key, event_id = row
-                    key = (intern_string(typ), intern_string(state_key))
-                    results[group][key] = event_id
+            overall_select_clause = " UNION ".join(select_clause_list)
         else:
```

Comment:
The diff is a bit weird to follow but we're just removing the SQLite specific `else` and unindenting

Reply:
FWIW https://github.com/matrix-org/synapse/pull/14531/files?w=1 shows it pretty clearly!